internal/contentmanager/dynamicsource.go

```diff
@@ -0,0 +1,183 @@
+	// received by the configured informer
+	Handler handler.EventHandler
+	// Predicates are the predicate.Predicate functions used to determine
+	// if a triggered event should be reacted to
+	Predicates []predicate.Predicate
+	// OnWatchError is the callback function that is used when the informer
+	// initially synced successfully and later encountered an error
+	OnWatchError func()
+	errOnce      sync.Once
+	errChan      chan error
+	syncedChan   chan struct{}
+	mu           sync.RWMutex
+	synced       bool
+	cancel       context.CancelFunc
+	ctx          context.Context
+	informer     informers.GenericInformer
+}
+
+var _ ReaderCloserSyncingSource = (*DynamicInformerSource)(nil)
+
+func (dis *DynamicInformerSource) String() string {
+	return "contentmanager.DynamicInformerSource"
+}
+
+func (dis *DynamicInformerSource) Start(ctx context.Context, q workqueue.RateLimitingInterface) error {
+	informerCtx, cancel := context.WithCancel(ctx)
+
+	dis.mu.RLock()
+	gInf := dis.DynamicInformerFactory.ForResource(dis.GVR)
+	dis.mu.RUnlock()
+
+	dis.mu.Lock()
+	dis.errChan = make(chan error, 1)
+	dis.syncedChan = make(chan struct{}, 1)
+	dis.informer = gInf
+	dis.cancel = cancel
+	dis.ctx = informerCtx
+	eventHandler := NewEventHandler[client.Object](ctx, q, dis.Handler, dis.Predicates)
+	dis.mu.Unlock()
+
+	// If we encounter an error during the watch we will:
+	// - Capture the error
+	// - cancel the informer
+	// requeuing of the ClusterExtension should happen by the
+	// WaitForSync function returning an error
+	// Only if we have successfully synced in the past should we
+	// requeue the ClusterExtension
+	sharedIndexInf := gInf.Informer()
+	err := sharedIndexInf.SetWatchErrorHandler(func(r *cgocache.Reflector, err error) {
+		dis.mu.Lock()
+		dis.errOnce.Do(func() {
+			dis.errChan <- err
+		})
+		dis.cancel()
+
+		if dis.synced {
+			// We won't be able to update the ClusterExtension status
+			// conditions so instead force a requeue if we
+			// have previously synced and then errored
+			q.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: dis.Owner.GetName(),
+				},
+			})
+			dis.OnWatchError()
+		}
+		dis.mu.Unlock()
+		cgocache.DefaultWatchErrorHandler(r, err)
+	})
+	if err != nil {
+		return fmt.Errorf("setting WatchErrorHandler: %w", err)
+	}
+
+	_, err = sharedIndexInf.AddEventHandler(eventHandler.HandlerFuncs())
+	if err != nil {
+		return fmt.Errorf("adding event handler: %w", err)
+	}
+
+	go sharedIndexInf.Run(informerCtx.Done())
+
+	go func() {
+		syncOnce := sync.OnceFunc(func() {
+			dis.mu.Lock()
+			dis.syncedChan <- struct{}{}
+			dis.synced = true
+			dis.mu.Unlock()
+		})
+
+		_ = wait.PollUntilContextCancel(informerCtx, time.Second, true, func(ctx context.Context) (bool, error) {
+			if sharedIndexInf.HasSynced() {
+				syncOnce()
+				return true, nil
+			}
+			return false, nil
+		})
+	}()
+
+	return nil
+}
+
+func (dis *DynamicInformerSource) WaitForSync(ctx context.Context) error {
```

Comment:
Add a check to see if we are already in an errored/synced state before waiting for channel responses